deeplearning4j-scaleout/spark/dl4j-spark-parameterserver/src/main/java/org/deeplearning4j/spark/parameterserver/pw/SharedTrainingWrapper.java

```diff
@@ -0,0 +1,340 @@
+ *
+ * @author raver119@gmail.com
+ */
+@Slf4j
+public class SharedTrainingWrapper {
+    public static SharedTrainingWrapper INSTANCE = new SharedTrainingWrapper();
+    protected ParallelWrapper wrapper;
+    protected VirtualDataSetIterator iteratorDS;
+    protected VirtualMultiDataSetIterator iteratorMDS;
+
+    protected List<Iterator<DataSet>> iteratorsDS;
+    protected List<Iterator<MultiDataSet>> iteratorsMDS;
+
+
+    protected AtomicBoolean isFirst = new AtomicBoolean(false);
+
+    protected ThreadLocal<BlockingObserver> observer = new ThreadLocal<>();
+    protected CudaGradientsAccumulator accumulator;
+
+    protected SilentTrainingDriver driver;
+
+    protected SharedTrainingWrapper() {
+        init();
+    }
+
+    protected void init() {
+        // instantiate some stuff here
+        iteratorsDS = new CopyOnWriteArrayList<>();
+        iteratorsMDS = new CopyOnWriteArrayList<>();
+
+        // now we're creating DataSetIterators, to feed ParallelWrapper
+        iteratorDS = new VirtualDataSetIterator(iteratorsDS);
+    }
+
+    public static SharedTrainingWrapper getInstance() {
+        return INSTANCE;
+    }
+
+    /**
+     * This method registers given Iterable<DataSet> in VirtualDataSetIterator
+     *
+     * @param iterator
+     */
+    public void attachDS(Iterator<DataSet> iterator) {
+        log.info("Attaching thread...");
+
+        // we're creating our Observable wrapper
+        VirtualIterator<DataSet> wrapped = new VirtualIterator<>(iterator);
+
+        // and creating Observer which will be used to monitor progress within iterator
+        BlockingObserver obs = new BlockingObserver();
+        wrapped.addObserver(obs);
+
+        // putting that "somewhere"
+        iteratorsDS.add(wrapped);
+
+        // storing observer into ThreadLocal, since we're going to use that later
+        observer.set(obs);
+    }
+
+    /**
+     * This method registers given Iterable<MultiDataSet> in VirtualMultiDataSetIterator
+     *
+     * @param iterator
+     */
+    public void attachMDS(Iterator<MultiDataSet> iterator) {
+        log.info("Attaching thread...");
+
+        // we're creating our Observable wrapper
+        VirtualIterator<MultiDataSet> wrapped = new VirtualIterator<>(iterator);
+
+        // and creating Observer which will be used to monitor progress within iterator
+        BlockingObserver obs = new BlockingObserver();
+        wrapped.addObserver(obs);
+
+        // putting that "somewhere"
+        iteratorsMDS.add(wrapped);
+
+        // storing observer into ThreadLocal, since we're going to use that later
+        observer.set(obs);
+    }
+
+    public SharedTrainingResult run(SharedTrainingWorker worker) {
+        /*
+            first call instantiates pw, messenger etc, and gets in charge here.
+         */
+        if (isFirst.compareAndSet(false, true)) {
+            // getting model from worker, and instantiating PW
+            try {
+                Thread.sleep(5000);
```

Comment:
Nah, you're right, that shouldn't get into master. I'll change mechanics from "First one runs PW" to "Last one runs PW"